src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -0,0 +1,324 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// This type provides means for instrumenting compiled methods for dynamic analysis.
+    /// It can be combined with other <see cref="Instrumenter"/>s.
+    /// </summary>
+    internal partial class DynamicAnalysisInjector : CompoundInstrumenter
+    {
+        private readonly bool _doInstrumentation;
+        private readonly MethodSymbol _method;
+        private readonly MethodSymbol _createPayload;
+        private readonly ArrayBuilder<SourceSpan> _spansBuilder;
+        private readonly ArrayTypeSymbol _payloadType;
+        private readonly LocalSymbol _methodPayload;
+        private readonly TypeCompilationState _compilationState;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly DebugDocumentProvider _debugDocumentProvider;
+        private readonly bool _methodHasExplicitBlock;
+        private readonly SyntheticBoundNodeFactory _factory;
+
+        public DynamicAnalysisInjector(MethodSymbol method, BoundStatement methodBody, SyntheticBoundNodeFactory factory, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, Instrumenter previous)
+            : base (previous)
+        {
+            CSharpCompilation compilation = compilationState.Compilation;
```

Comment:
I'm reluctant to introduce "var" here. I use it only when editing code that uses already uses it.